File: floyd.lua
n,e = io.read("*n", "*n")
edges = {}
for i = 1,n do
  edges[i] = {}
  for j = 1,n do
    if (i == j) then edges[i][j] = 0 else edges[i][j] = -1 end
  end
end

for i = 1,e do
  n1,n2,w = io.read("*n", "*n", "*n")
  edges[n1][n2] = w
end

for k = 1,n do
  for i = 1,n do
    if (edges[i][k] ~= -1) then
      for j = 1,n do
        if (edges[k][j] ~= -1 and (edges[i][j] == -1 or edges[i][k] + edges[k][j] < edges[i][j])) then
          edges[i][j] = edges[i][k] + edges[k][j]
        end
      end
    end
  end
end

tests = io.read("*n")
for i = 1,tests do
  nt1,nt2 = io.read("*n", "*n")
  print(edges[nt1][nt2])
end

File: floyd.ml
open Str

let [n; e] = read_line() |> Str.(split (regexp " ")) |> List.map int_of_string

let edges = Array.init n (fun _ -> Array.make n ~-1)

let _ =
  for _ = 1 to e do
    let [n1; n2; w] = read_line() |> Str.(split (regexp " ")) |> List.map int_of_string in
    edges.(n1-1).(n2-1) <- w
  done;

  for k = 0 to n-1 do
    for i = 0 to n-1 do
      if edges.(i).(k) <> ~-1 then
        for j = 0 to n-1 do
          if edges.(k).(j) <> ~-1 && (edges.(i).(j) = ~-1 || edges.(i).(k) + edges.(k).(j) < edges.(i).(j)) then
            edges.(i).(j) <- edges.(i).(k) + edges.(k).(j)
        done
    done
  done;

  let tests = read_line() |> int_of_string in

  for _ = 1 to tests do
    let [nt1; nt2] = read_line() |> Str.(split (regexp " ")) |> List.map int_of_string in
    edges.(nt1-1).(nt2-1) |> print_int;
    print_newline()
  done
